src/main/scala/io/iohk/ethereum/vm/OpCode.scala

```diff
@@ -0,0 +1,507 @@
+    val updatedState = for {
+      popped <- state.stack.pop(4)
+      (Seq(address, memOffset, codeOffset, size), stack1) = popped
+      codeCopy: ByteString = ??? //TODO: program.getCode(address).drop(codeOffset)
+      mem1 = state.memory.store(memOffset, codeCopy)
+    } yield state.withStack(stack1).withMemory(mem1).step()
+
+    updatedState.valueOr(state.withError)
+  }
+}
+
+case object POP extends OpCode(0x50, 1, 0) {
+  def execute(state: ProgramState): ProgramState = {
+    state.stack.pop
+      .map { case (_, stack) => state.withStack(stack).step() }
+      .valueOr(state.withError)
+  }
+}
+
+case object MLOAD extends OpCode(0x51, 1, 1) {
+  def execute(state: ProgramState): ProgramState = {
+    val updatedState = for {
+      popped <- state.stack.pop
+      (addr, stack1) = popped
+      (word, mem1) = state.memory.load(addr)
+      stack2 <- stack1.push(word)
+    } yield state.withStack(stack2).withMemory(mem1).step()
+
+    updatedState.valueOr(state.withError)
+  }
+}
+
+case object MSTORE extends OpCode(0x52, 2, 0) {
+  def execute(state: ProgramState): ProgramState = {
+    val updatedState = for {
+      popped <- state.stack.pop(2)
+      (Seq(addr, value), stack1) = popped
+      updatedMem = state.memory.store(addr, value)
+    } yield state.withStack(stack1).withMemory(updatedMem).step()
+
+    updatedState.valueOr(state.withError)
+  }
+}
+
+case object SLOAD extends OpCode(0x54, 1, 1) {
+  def execute(state: ProgramState): ProgramState = {
+    val updatedState = for {
+      popped <- state.stack.pop
+      (addr, stack1) = popped
+      value = state.storage.load(addr)
+      stack2 <- stack1.push(value)
+    } yield state.withStack(stack2).step()
+
+    updatedState.valueOr(state.withError)
+  }
+}
+
+
+case object SSTORE extends OpCode(0x55, 2, 0) {
+  def execute(state: ProgramState): ProgramState = {
+    val updatedState = for {
+      popped <- state.stack.pop(2)
+      (Seq(addr, value), stack1) = popped
+      updatedStorage = state.storage.store(addr, value)
+    } yield state.withStack(stack1).withStorage(updatedStorage).step()
+
+    updatedState.valueOr(state.withError)
+  }
+}
+
+
+case object JUMP extends OpCode(0x56, 1, 0) {
+  def execute(state: ProgramState): ProgramState = {
+    val updatedState = for {
+      popped <- state.stack.pop
+      (pos, stack1) = popped
+    } yield state.withStack(stack1).goto(pos.intValue)
+
+    updatedState.valueOr(state.withError)
+  }
+}
+
+case object JUMPI extends OpCode(0x57, 2, 0) {
+  def execute(state: ProgramState): ProgramState = {
+    val updatedState = for {
+      popped <- state.stack.pop(2)
+      (Seq(pos, cond), stack1) = popped
+      nextPos = if (cond != 0) pos.intValue else state.pc + 1
+    } yield state.withStack(stack1).goto(nextPos)
+
+    updatedState.valueOr(state.withError)
+  }
+}
+
+case object JUMPDEST extends OpCode(0x5b, 0, 0) {
+  def execute(state: ProgramState): ProgramState = {
+    //TODO: what is it for, really?
```

Comment:
Yes, we actually already know what it is. This is just a bit inaccurate comment to remember to do the verification. But it will be a separate PR. For now I can update the comment.